src/Symfony/Component/EventDispatcher/EventSubscriberInterface.php

```diff
@@ -40,7 +40,7 @@ interface EventSubscriberInterface
      *  * ['eventName' => ['methodName', $priority]]
      *  * ['eventName' => [['methodName1', $priority], ['methodName2']]]
      *
-     * @return array The event names to listen to
+     * @return string[]|string[][]|int[][] The event names to listen to
```

Comment:
I'm not an expert in phpdocs, but it looks wrong to me. Something like `map[string]string` describes the first use case, and then it gets more complex :)

Reply:
Maybe you are confused because you try to model keys too? Official PHPDoc have no way to document these, it's values only. 

So first use case is string[] because it's array of string values.
Second use case is string[][]|int[][] because it's array of arrays with string or int values.

Reply:
I fail to see how this is useful then. As you can have a mix of strings and ints, shouldn't it be something like `(string|int)[][]`?

Reply:
(string|int)[][] and string[][]|int[][] is identical (except latter is more standard and has better support). It's always union, there is no support for specifying array must contain both

Reply:
Are you sure they are identical? We are already using the first form in Symfony.

Reply:
@muglug @ondrejmirtes can you chip in?

Reply:
They are not exactly identical but AFAIK the current static analysis tools do not make any difference between them.

The first form (`(string|int)[]`) - each array item can be either a string or an int
The second form (`string[]|int[]`) - It's either an array of strings or an array of integers. So by looking at the type of the first item, you'd be able to infer how the rest of the array looks like.

Anyway, this return typehint `string[]|string[][]|int[][]` is already pretty complex and unusable without an advanced dynamic return type extension that will determine what actually gets returned for each specific call, right?

Reply:
I understand the general motivation behind refining the typehint, but I don't see any real improvement in this exact case. New type gives almost no additional information about the return type. Neither for humans, nor for machines. You still have to check description for details to implement or use the interface correctly.

@ostrolucky , does any static analyzer complain about the existing phpdoc? In other words, what exact problem this change solves?

Reply:
> In other words, what exact problem this change solves?

Psalm complains on its strictest setting when it can't infer a return type: https://getpsalm.org/r/4fe8fcdbef

If it were me I'd probably use a stub file with a much more specific return type (supported by both Psalm and Phan, but not currently PHPStan):
`array<string, string|array{string, int}|array<int, array{string,int}>>`
(example in use here: https://getpsalm.org/r/a70f39dc6a)

this informs the consumer that it either returns an array of strings, an array of `{string, int}` tuple arrays, or an array of arrays of `{string, int}` tuples, all keyed by a string.

Reply:
> New type gives almost no additional information about the return type.

Sure it does. SA tool can infer it's an error having int[] or having something else than integers or strings in an array, or when it's more dimensional than that.

> what exact problem this change solves?

There are enforced style guides in projects which dictate every array must specify type of its members. By implementing interface without such docblock, IDE won't create docblock matching such requirements. This change is simple way how to solve this problem and additionally have more type safety thanks to SA tools.

> Anyway, this return typehint `string[]|string[][]|int[][]` is already pretty complex

Yes, but it matches how this method works. It's not a fault of docblock that method accepts complex types, it reflects it.

> I'd probably use a stub file

That would be fine when you are tied to Psalm, but there is zero support for such syntax outside of it. That's why I wrote there is no support for specifying array must contain both types. PHPDocumentor doesn't have such thing and that's the only official thing we have, so as far as I see what I used here is best we got now.

Reply:
Given that `string[]|string[][]|int[][]` is complex to read while still not accurately describing the return type (well, actually describing a wrong one), I don't think it make sense to use it.

Until we have a common agreement on the way to document a type like that (maybe the one Psalm and Phan support today, which indeed describes the right type), I don't think it makes sense to change the phpdoc in this interface.

Reply:
> The first form (`(string|int)[]`) - each array item can be either a string or an int
The second form (`string[]|int[]`) - It's either an array of strings or an array of integers.

@ostrolucky , okay, then based on @ondrejmirtes comment we should pick `(string|int)[][]` as @fabpot proposed. It's true that each item of the subarray is either a string or an int, but it's false that method ever returns an array of arrays of ints.

Reply:
If I'm not mistaken, the correct type would be `(string|(string|int)[]|(string|int)[][])[]` which is arguably hard to read.